apex/contrib/csrc/nccl_allocator/NCCLAllocator.cpp

```diff
@@ -0,0 +1,437 @@
+    if (result != ncclSuccess) {                                          \
+      std::string err = "NCCL error in: " + std::string(__FILE__) + ":" + \
+          std::to_string(__LINE__) + ", " +                               \
+          std::string(ncclGetErrorString(result));                        \
+      TORCH_CHECK(false, err);                                            \
+    }                                                                     \
+  } while (0)
+
+namespace nccl_allocator::cuda {
+
+extern int device_count;
+
+extern thread_local bool _use_nccl_mem;
+
+std::shared_ptr<c10::cuda::CUDACachingAllocator::CUDAAllocator>
+createNCCLAllocator() {
+  auto allocator = std::make_shared<NCCLAllocator>();
+  allocator->init(device_count);
+  return allocator;
+}
+
+void custom_raw_deleter(void* ptr);
+
+_AllocationMetadata::_AllocationMetadata()
+    : size(0), device_idx(-1), stream{}, ptr(nullptr), is_free(false) {}
+
+_AllocationMetadata::_AllocationMetadata(
+    size_t size,
+    c10::DeviceIndex device_idx,
+    cudaStream_t stream,
+    void* ptr)
+    : size(size), device_idx(device_idx), stream(stream), ptr(ptr), is_free(false) {}
+
+NCCLAllocator::NCCLAllocator(NCCLAllocator& other)
+    : init_fn_(other.init_fn_),
+      memory_fraction_fn_(other.memory_fraction_fn_),
+      base_alloc_fn_(other.base_alloc_fn_),
+      record_stream_fn_(other.record_stream_fn_),
+      begin_allocate_to_pool_fn_(other.begin_allocate_to_pool_fn_),
+      end_allocate_to_pool_fn_(other.end_allocate_to_pool_fn_),
+      relase_pool_fn_(other.relase_pool_fn_) {}
+
+void* NCCLAllocator::malloc(size_t size, c10::DeviceIndex device, cudaStream_t stream) {
+  void* r = nullptr;
+#ifdef NCCL_SUPPORTS_UB
+  if (_use_nccl_mem) {
+    for (
+      auto md = nccl_allocation_metadata_.begin();
+           md != nccl_allocation_metadata_.end();
+	   md++
+    ) {
+      // compare stream?
+      if(md->second.is_free && md->second.size == size && md->second.device_idx == device) {
+        md->second.is_free = false;
+	r = md->second.ptr;
+	return r;
+      }
+    }
+
+    C10_NCCL_CHECK(ncclMemAlloc(&r, size));
+    {
+      const std::lock_guard<std::mutex> lock(allocator_mutex_);
+      nccl_allocation_metadata_.emplace(r, _AllocationMetadata(size, device, stream, r));
+
+      auto te = c10::cuda::CUDACachingAllocator::TraceEntry(
+        c10::cuda::CUDACachingAllocator::TraceEntry::Action::SEGMENT_ALLOC,
+        device,
+        (int64_t)r,
+        size,
+        stream,
+        c10::getApproximateTime(),
+        nullptr);
+      for (const auto& cb : trace_trackers_) {
+        cb(te);
+      }
+    }
+  }
+#else
+  if (_use_nccl_mem) {
+    TORCH_WARN_ONCE("ncclMemAlloc is not supported. ",
+                    "NCCL v2.19 and newer is required,",
+                    "but found NCCL v", NCCL_MAJOR, ".", NCCL_MINOR, ". ",
+                    "Falling back to cudaMalloc.");
+  }
+  if (false) {
+    // goto cudaMalloc
+  }
```

Comment:
is this... approached?

Reply:
Potentially yes, if someone tries to compile it with NCCL older than 2.19.